// Generated by the protocol buffer compiler.  DO NOT EDIT!
// source: VcfRecord.proto

package org.opencb.biodata.models.variant.protobuf;

public final class VcfSliceProtos {
  private VcfSliceProtos() {}
  public static void registerAllExtensions(
      com.google.protobuf.ExtensionRegistry registry) {
  }
  public interface VcfSampleOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VcfSample)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>repeated string sample_values = 1;</code>
     */
    com.google.protobuf.ProtocolStringList
        getSampleValuesList();
    /**
     * <code>repeated string sample_values = 1;</code>
     */
    int getSampleValuesCount();
    /**
     * <code>repeated string sample_values = 1;</code>
     */
    java.lang.String getSampleValues(int index);
    /**
     * <code>repeated string sample_values = 1;</code>
     */
    com.google.protobuf.ByteString
        getSampleValuesBytes(int index);

    /**
     * <code>optional uint32 gt_index = 2;</code>
     *
     * <pre>
     * GT is mandatory. Saving it separately can create a map of genotypes in Fields
     * </pre>
     */
    int getGtIndex();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VcfSample}
   */
  public  static final class VcfSample extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VcfSample)
      VcfSampleOrBuilder {
    // Use VcfSample.newBuilder() to construct.
    private VcfSample(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private VcfSample() {
      sampleValues_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      gtIndex_ = 0;
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private VcfSample(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
                sampleValues_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000001;
              }
              sampleValues_.add(s);
              break;
            }
            case 16: {

              gtIndex_ = input.readUInt32();
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
          sampleValues_ = sampleValues_.getUnmodifiableView();
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder.class);
    }

    private int bitField0_;
    public static final int SAMPLE_VALUES_FIELD_NUMBER = 1;
    private com.google.protobuf.LazyStringList sampleValues_;
    /**
     * <code>repeated string sample_values = 1;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getSampleValuesList() {
      return sampleValues_;
    }
    /**
     * <code>repeated string sample_values = 1;</code>
     */
    public int getSampleValuesCount() {
      return sampleValues_.size();
    }
    /**
     * <code>repeated string sample_values = 1;</code>
     */
    public java.lang.String getSampleValues(int index) {
      return sampleValues_.get(index);
    }
    /**
     * <code>repeated string sample_values = 1;</code>
     */
    public com.google.protobuf.ByteString
        getSampleValuesBytes(int index) {
      return sampleValues_.getByteString(index);
    }

    public static final int GT_INDEX_FIELD_NUMBER = 2;
    private int gtIndex_;
    /**
     * <code>optional uint32 gt_index = 2;</code>
     *
     * <pre>
     * GT is mandatory. Saving it separately can create a map of genotypes in Fields
     * </pre>
     */
    public int getGtIndex() {
      return gtIndex_;
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      for (int i = 0; i < sampleValues_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, sampleValues_.getRaw(i));
      }
      if (gtIndex_ != 0) {
        output.writeUInt32(2, gtIndex_);
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      {
        int dataSize = 0;
        for (int i = 0; i < sampleValues_.size(); i++) {
          dataSize += computeStringSizeNoTag(sampleValues_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getSampleValuesList().size();
      }
      if (gtIndex_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeUInt32Size(2, gtIndex_);
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VcfSample}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VcfSample)
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        sampleValues_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000001);
        gtIndex_ = 0;

        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample build() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample result = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        if (((bitField0_ & 0x00000001) == 0x00000001)) {
          sampleValues_ = sampleValues_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000001);
        }
        result.sampleValues_ = sampleValues_;
        result.gtIndex_ = gtIndex_;
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.getDefaultInstance()) return this;
        if (!other.sampleValues_.isEmpty()) {
          if (sampleValues_.isEmpty()) {
            sampleValues_ = other.sampleValues_;
            bitField0_ = (bitField0_ & ~0x00000001);
          } else {
            ensureSampleValuesIsMutable();
            sampleValues_.addAll(other.sampleValues_);
          }
          onChanged();
        }
        if (other.getGtIndex() != 0) {
          setGtIndex(other.getGtIndex());
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private com.google.protobuf.LazyStringList sampleValues_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureSampleValuesIsMutable() {
        if (!((bitField0_ & 0x00000001) == 0x00000001)) {
          sampleValues_ = new com.google.protobuf.LazyStringArrayList(sampleValues_);
          bitField0_ |= 0x00000001;
         }
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getSampleValuesList() {
        return sampleValues_.getUnmodifiableView();
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public int getSampleValuesCount() {
        return sampleValues_.size();
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public java.lang.String getSampleValues(int index) {
        return sampleValues_.get(index);
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public com.google.protobuf.ByteString
          getSampleValuesBytes(int index) {
        return sampleValues_.getByteString(index);
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public Builder setSampleValues(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureSampleValuesIsMutable();
        sampleValues_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public Builder addSampleValues(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureSampleValuesIsMutable();
        sampleValues_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public Builder addAllSampleValues(
          java.lang.Iterable<java.lang.String> values) {
        ensureSampleValuesIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, sampleValues_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public Builder clearSampleValues() {
        sampleValues_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000001);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string sample_values = 1;</code>
       */
      public Builder addSampleValuesBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureSampleValuesIsMutable();
        sampleValues_.add(value);
        onChanged();
        return this;
      }

      private int gtIndex_ ;
      /**
       * <code>optional uint32 gt_index = 2;</code>
       *
       * <pre>
       * GT is mandatory. Saving it separately can create a map of genotypes in Fields
       * </pre>
       */
      public int getGtIndex() {
        return gtIndex_;
      }
      /**
       * <code>optional uint32 gt_index = 2;</code>
       *
       * <pre>
       * GT is mandatory. Saving it separately can create a map of genotypes in Fields
       * </pre>
       */
      public Builder setGtIndex(int value) {
        
        gtIndex_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional uint32 gt_index = 2;</code>
       *
       * <pre>
       * GT is mandatory. Saving it separately can create a map of genotypes in Fields
       * </pre>
       */
      public Builder clearGtIndex() {
        
        gtIndex_ = 0;
        onChanged();
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VcfSample)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VcfSample)
    private static final org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample();
    }

    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<VcfSample>
        PARSER = new com.google.protobuf.AbstractParser<VcfSample>() {
      public VcfSample parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new VcfSample(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<VcfSample> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<VcfSample> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface VcfRecordOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VcfRecord)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional int32 relative_start = 1;</code>
     *
     * <pre>
     * 1 based
     * May contain negative values but it's not likely
     * </pre>
     */
    int getRelativeStart();

    /**
     * <code>optional int32 relative_end = 2;</code>
     *
     * <pre>
     * May contain negative values but it's not likely
     * </pre>
     */
    int getRelativeEnd();

    /**
     * <code>optional string reference = 3;</code>
     */
    java.lang.String getReference();
    /**
     * <code>optional string reference = 3;</code>
     */
    com.google.protobuf.ByteString
        getReferenceBytes();

    /**
     * <code>optional string alternate = 4;</code>
     */
    java.lang.String getAlternate();
    /**
     * <code>optional string alternate = 4;</code>
     */
    com.google.protobuf.ByteString
        getAlternateBytes();

    /**
     * <code>optional float quality = 5;</code>
     */
    float getQuality();

    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
     */
    int getTypeValue();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getType();

    /**
     * <code>optional string call = 13;</code>
     */
    java.lang.String getCall();
    /**
     * <code>optional string call = 13;</code>
     */
    com.google.protobuf.ByteString
        getCallBytes();

    /**
     * <code>optional uint32 filter_index = 6;</code>
     */
    int getFilterIndex();

    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    com.google.protobuf.ProtocolStringList
        getIdNonDefaultList();
    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    int getIdNonDefaultCount();
    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    java.lang.String getIdNonDefault(int index);
    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    com.google.protobuf.ByteString
        getIdNonDefaultBytes(int index);

    /**
     * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
     */
    java.util.List<java.lang.Integer> getInfoKeyIndexList();
    /**
     * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
     */
    int getInfoKeyIndexCount();
    /**
     * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
     */
    int getInfoKeyIndex(int index);

    /**
     * <code>repeated string info_value = 9;</code>
     */
    com.google.protobuf.ProtocolStringList
        getInfoValueList();
    /**
     * <code>repeated string info_value = 9;</code>
     */
    int getInfoValueCount();
    /**
     * <code>repeated string info_value = 9;</code>
     */
    java.lang.String getInfoValue(int index);
    /**
     * <code>repeated string info_value = 9;</code>
     */
    com.google.protobuf.ByteString
        getInfoValueBytes(int index);

    /**
     * <code>optional uint32 formatIndex = 10;</code>
     */
    int getFormatIndex();

    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample> 
        getSamplesList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample getSamples(int index);
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    int getSamplesCount();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder> 
        getSamplesOrBuilderList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder getSamplesOrBuilder(
        int index);

    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate> 
        getSecondaryAlternatesList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate getSecondaryAlternates(int index);
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    int getSecondaryAlternatesCount();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder> 
        getSecondaryAlternatesOrBuilderList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder getSecondaryAlternatesOrBuilder(
        int index);
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VcfRecord}
   */
  public  static final class VcfRecord extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VcfRecord)
      VcfRecordOrBuilder {
    // Use VcfRecord.newBuilder() to construct.
    private VcfRecord(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private VcfRecord() {
      relativeStart_ = 0;
      relativeEnd_ = 0;
      reference_ = "";
      alternate_ = "";
      quality_ = 0F;
      type_ = 0;
      call_ = "";
      filterIndex_ = 0;
      idNonDefault_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      infoKeyIndex_ = java.util.Collections.emptyList();
      infoValue_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      formatIndex_ = 0;
      samples_ = java.util.Collections.emptyList();
      secondaryAlternates_ = java.util.Collections.emptyList();
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private VcfRecord(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 8: {

              relativeStart_ = input.readInt32();
              break;
            }
            case 16: {

              relativeEnd_ = input.readInt32();
              break;
            }
            case 26: {
              String s = input.readStringRequireUtf8();

              reference_ = s;
              break;
            }
            case 34: {
              String s = input.readStringRequireUtf8();

              alternate_ = s;
              break;
            }
            case 45: {

              quality_ = input.readFloat();
              break;
            }
            case 48: {

              filterIndex_ = input.readUInt32();
              break;
            }
            case 58: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000100) == 0x00000100)) {
                idNonDefault_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000100;
              }
              idNonDefault_.add(s);
              break;
            }
            case 64: {
              if (!((mutable_bitField0_ & 0x00000200) == 0x00000200)) {
                infoKeyIndex_ = new java.util.ArrayList<java.lang.Integer>();
                mutable_bitField0_ |= 0x00000200;
              }
              infoKeyIndex_.add(input.readUInt32());
              break;
            }
            case 66: {
              int length = input.readRawVarint32();
              int limit = input.pushLimit(length);
              if (!((mutable_bitField0_ & 0x00000200) == 0x00000200) && input.getBytesUntilLimit() > 0) {
                infoKeyIndex_ = new java.util.ArrayList<java.lang.Integer>();
                mutable_bitField0_ |= 0x00000200;
              }
              while (input.getBytesUntilLimit() > 0) {
                infoKeyIndex_.add(input.readUInt32());
              }
              input.popLimit(limit);
              break;
            }
            case 74: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000400) == 0x00000400)) {
                infoValue_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000400;
              }
              infoValue_.add(s);
              break;
            }
            case 80: {

              formatIndex_ = input.readUInt32();
              break;
            }
            case 90: {
              if (!((mutable_bitField0_ & 0x00001000) == 0x00001000)) {
                samples_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample>();
                mutable_bitField0_ |= 0x00001000;
              }
              samples_.add(input.readMessage(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.parser(), extensionRegistry));
              break;
            }
            case 96: {
              int rawValue = input.readEnum();

              type_ = rawValue;
              break;
            }
            case 106: {
              String s = input.readStringRequireUtf8();

              call_ = s;
              break;
            }
            case 114: {
              if (!((mutable_bitField0_ & 0x00002000) == 0x00002000)) {
                secondaryAlternates_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate>();
                mutable_bitField0_ |= 0x00002000;
              }
              secondaryAlternates_.add(input.readMessage(org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.parser(), extensionRegistry));
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000100) == 0x00000100)) {
          idNonDefault_ = idNonDefault_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00000200) == 0x00000200)) {
          infoKeyIndex_ = java.util.Collections.unmodifiableList(infoKeyIndex_);
        }
        if (((mutable_bitField0_ & 0x00000400) == 0x00000400)) {
          infoValue_ = infoValue_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00001000) == 0x00001000)) {
          samples_ = java.util.Collections.unmodifiableList(samples_);
        }
        if (((mutable_bitField0_ & 0x00002000) == 0x00002000)) {
          secondaryAlternates_ = java.util.Collections.unmodifiableList(secondaryAlternates_);
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder.class);
    }

    private int bitField0_;
    public static final int RELATIVE_START_FIELD_NUMBER = 1;
    private int relativeStart_;
    /**
     * <code>optional int32 relative_start = 1;</code>
     *
     * <pre>
     * 1 based
     * May contain negative values but it's not likely
     * </pre>
     */
    public int getRelativeStart() {
      return relativeStart_;
    }

    public static final int RELATIVE_END_FIELD_NUMBER = 2;
    private int relativeEnd_;
    /**
     * <code>optional int32 relative_end = 2;</code>
     *
     * <pre>
     * May contain negative values but it's not likely
     * </pre>
     */
    public int getRelativeEnd() {
      return relativeEnd_;
    }

    public static final int REFERENCE_FIELD_NUMBER = 3;
    private volatile java.lang.Object reference_;
    /**
     * <code>optional string reference = 3;</code>
     */
    public java.lang.String getReference() {
      java.lang.Object ref = reference_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        reference_ = s;
        return s;
      }
    }
    /**
     * <code>optional string reference = 3;</code>
     */
    public com.google.protobuf.ByteString
        getReferenceBytes() {
      java.lang.Object ref = reference_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        reference_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int ALTERNATE_FIELD_NUMBER = 4;
    private volatile java.lang.Object alternate_;
    /**
     * <code>optional string alternate = 4;</code>
     */
    public java.lang.String getAlternate() {
      java.lang.Object ref = alternate_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        alternate_ = s;
        return s;
      }
    }
    /**
     * <code>optional string alternate = 4;</code>
     */
    public com.google.protobuf.ByteString
        getAlternateBytes() {
      java.lang.Object ref = alternate_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        alternate_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int QUALITY_FIELD_NUMBER = 5;
    private float quality_;
    /**
     * <code>optional float quality = 5;</code>
     */
    public float getQuality() {
      return quality_;
    }

    public static final int TYPE_FIELD_NUMBER = 12;
    private int type_;
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
     */
    public int getTypeValue() {
      return type_;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getType() {
      org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType result = org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.valueOf(type_);
      return result == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.UNRECOGNIZED : result;
    }

    public static final int CALL_FIELD_NUMBER = 13;
    private volatile java.lang.Object call_;
    /**
     * <code>optional string call = 13;</code>
     */
    public java.lang.String getCall() {
      java.lang.Object ref = call_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        call_ = s;
        return s;
      }
    }
    /**
     * <code>optional string call = 13;</code>
     */
    public com.google.protobuf.ByteString
        getCallBytes() {
      java.lang.Object ref = call_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        call_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int FILTER_INDEX_FIELD_NUMBER = 6;
    private int filterIndex_;
    /**
     * <code>optional uint32 filter_index = 6;</code>
     */
    public int getFilterIndex() {
      return filterIndex_;
    }

    public static final int ID_NON_DEFAULT_FIELD_NUMBER = 7;
    private com.google.protobuf.LazyStringList idNonDefault_;
    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getIdNonDefaultList() {
      return idNonDefault_;
    }
    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    public int getIdNonDefaultCount() {
      return idNonDefault_.size();
    }
    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    public java.lang.String getIdNonDefault(int index) {
      return idNonDefault_.get(index);
    }
    /**
     * <code>repeated string id_non_default = 7;</code>
     */
    public com.google.protobuf.ByteString
        getIdNonDefaultBytes(int index) {
      return idNonDefault_.getByteString(index);
    }

    public static final int INFO_KEY_INDEX_FIELD_NUMBER = 8;
    private java.util.List<java.lang.Integer> infoKeyIndex_;
    /**
     * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
     */
    public java.util.List<java.lang.Integer>
        getInfoKeyIndexList() {
      return infoKeyIndex_;
    }
    /**
     * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
     */
    public int getInfoKeyIndexCount() {
      return infoKeyIndex_.size();
    }
    /**
     * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
     */
    public int getInfoKeyIndex(int index) {
      return infoKeyIndex_.get(index);
    }
    private int infoKeyIndexMemoizedSerializedSize = -1;

    public static final int INFO_VALUE_FIELD_NUMBER = 9;
    private com.google.protobuf.LazyStringList infoValue_;
    /**
     * <code>repeated string info_value = 9;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getInfoValueList() {
      return infoValue_;
    }
    /**
     * <code>repeated string info_value = 9;</code>
     */
    public int getInfoValueCount() {
      return infoValue_.size();
    }
    /**
     * <code>repeated string info_value = 9;</code>
     */
    public java.lang.String getInfoValue(int index) {
      return infoValue_.get(index);
    }
    /**
     * <code>repeated string info_value = 9;</code>
     */
    public com.google.protobuf.ByteString
        getInfoValueBytes(int index) {
      return infoValue_.getByteString(index);
    }

    public static final int FORMATINDEX_FIELD_NUMBER = 10;
    private int formatIndex_;
    /**
     * <code>optional uint32 formatIndex = 10;</code>
     */
    public int getFormatIndex() {
      return formatIndex_;
    }

    public static final int SAMPLES_FIELD_NUMBER = 11;
    private java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample> samples_;
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    public java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample> getSamplesList() {
      return samples_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder> 
        getSamplesOrBuilderList() {
      return samples_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    public int getSamplesCount() {
      return samples_.size();
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample getSamples(int index) {
      return samples_.get(index);
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder getSamplesOrBuilder(
        int index) {
      return samples_.get(index);
    }

    public static final int SECONDARYALTERNATES_FIELD_NUMBER = 14;
    private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate> secondaryAlternates_;
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate> getSecondaryAlternatesList() {
      return secondaryAlternates_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder> 
        getSecondaryAlternatesOrBuilderList() {
      return secondaryAlternates_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    public int getSecondaryAlternatesCount() {
      return secondaryAlternates_.size();
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate getSecondaryAlternates(int index) {
      return secondaryAlternates_.get(index);
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder getSecondaryAlternatesOrBuilder(
        int index) {
      return secondaryAlternates_.get(index);
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      getSerializedSize();
      if (relativeStart_ != 0) {
        output.writeInt32(1, relativeStart_);
      }
      if (relativeEnd_ != 0) {
        output.writeInt32(2, relativeEnd_);
      }
      if (!getReferenceBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 3, reference_);
      }
      if (!getAlternateBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 4, alternate_);
      }
      if (quality_ != 0F) {
        output.writeFloat(5, quality_);
      }
      if (filterIndex_ != 0) {
        output.writeUInt32(6, filterIndex_);
      }
      for (int i = 0; i < idNonDefault_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 7, idNonDefault_.getRaw(i));
      }
      if (getInfoKeyIndexList().size() > 0) {
        output.writeRawVarint32(66);
        output.writeRawVarint32(infoKeyIndexMemoizedSerializedSize);
      }
      for (int i = 0; i < infoKeyIndex_.size(); i++) {
        output.writeUInt32NoTag(infoKeyIndex_.get(i));
      }
      for (int i = 0; i < infoValue_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 9, infoValue_.getRaw(i));
      }
      if (formatIndex_ != 0) {
        output.writeUInt32(10, formatIndex_);
      }
      for (int i = 0; i < samples_.size(); i++) {
        output.writeMessage(11, samples_.get(i));
      }
      if (type_ != org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.NO_VARIATION.getNumber()) {
        output.writeEnum(12, type_);
      }
      if (!getCallBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 13, call_);
      }
      for (int i = 0; i < secondaryAlternates_.size(); i++) {
        output.writeMessage(14, secondaryAlternates_.get(i));
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (relativeStart_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(1, relativeStart_);
      }
      if (relativeEnd_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeInt32Size(2, relativeEnd_);
      }
      if (!getReferenceBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(3, reference_);
      }
      if (!getAlternateBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(4, alternate_);
      }
      if (quality_ != 0F) {
        size += com.google.protobuf.CodedOutputStream
          .computeFloatSize(5, quality_);
      }
      if (filterIndex_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeUInt32Size(6, filterIndex_);
      }
      {
        int dataSize = 0;
        for (int i = 0; i < idNonDefault_.size(); i++) {
          dataSize += computeStringSizeNoTag(idNonDefault_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getIdNonDefaultList().size();
      }
      {
        int dataSize = 0;
        for (int i = 0; i < infoKeyIndex_.size(); i++) {
          dataSize += com.google.protobuf.CodedOutputStream
            .computeUInt32SizeNoTag(infoKeyIndex_.get(i));
        }
        size += dataSize;
        if (!getInfoKeyIndexList().isEmpty()) {
          size += 1;
          size += com.google.protobuf.CodedOutputStream
              .computeInt32SizeNoTag(dataSize);
        }
        infoKeyIndexMemoizedSerializedSize = dataSize;
      }
      {
        int dataSize = 0;
        for (int i = 0; i < infoValue_.size(); i++) {
          dataSize += computeStringSizeNoTag(infoValue_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getInfoValueList().size();
      }
      if (formatIndex_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeUInt32Size(10, formatIndex_);
      }
      for (int i = 0; i < samples_.size(); i++) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(11, samples_.get(i));
      }
      if (type_ != org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.NO_VARIATION.getNumber()) {
        size += com.google.protobuf.CodedOutputStream
          .computeEnumSize(12, type_);
      }
      if (!getCallBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(13, call_);
      }
      for (int i = 0; i < secondaryAlternates_.size(); i++) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(14, secondaryAlternates_.get(i));
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VcfRecord}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VcfRecord)
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
          getSamplesFieldBuilder();
          getSecondaryAlternatesFieldBuilder();
        }
      }
      public Builder clear() {
        super.clear();
        relativeStart_ = 0;

        relativeEnd_ = 0;

        reference_ = "";

        alternate_ = "";

        quality_ = 0F;

        type_ = 0;

        call_ = "";

        filterIndex_ = 0;

        idNonDefault_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000100);
        infoKeyIndex_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000200);
        infoValue_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000400);
        formatIndex_ = 0;

        if (samplesBuilder_ == null) {
          samples_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00001000);
        } else {
          samplesBuilder_.clear();
        }
        if (secondaryAlternatesBuilder_ == null) {
          secondaryAlternates_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00002000);
        } else {
          secondaryAlternatesBuilder_.clear();
        }
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord build() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord result = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.relativeStart_ = relativeStart_;
        result.relativeEnd_ = relativeEnd_;
        result.reference_ = reference_;
        result.alternate_ = alternate_;
        result.quality_ = quality_;
        result.type_ = type_;
        result.call_ = call_;
        result.filterIndex_ = filterIndex_;
        if (((bitField0_ & 0x00000100) == 0x00000100)) {
          idNonDefault_ = idNonDefault_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000100);
        }
        result.idNonDefault_ = idNonDefault_;
        if (((bitField0_ & 0x00000200) == 0x00000200)) {
          infoKeyIndex_ = java.util.Collections.unmodifiableList(infoKeyIndex_);
          bitField0_ = (bitField0_ & ~0x00000200);
        }
        result.infoKeyIndex_ = infoKeyIndex_;
        if (((bitField0_ & 0x00000400) == 0x00000400)) {
          infoValue_ = infoValue_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000400);
        }
        result.infoValue_ = infoValue_;
        result.formatIndex_ = formatIndex_;
        if (samplesBuilder_ == null) {
          if (((bitField0_ & 0x00001000) == 0x00001000)) {
            samples_ = java.util.Collections.unmodifiableList(samples_);
            bitField0_ = (bitField0_ & ~0x00001000);
          }
          result.samples_ = samples_;
        } else {
          result.samples_ = samplesBuilder_.build();
        }
        if (secondaryAlternatesBuilder_ == null) {
          if (((bitField0_ & 0x00002000) == 0x00002000)) {
            secondaryAlternates_ = java.util.Collections.unmodifiableList(secondaryAlternates_);
            bitField0_ = (bitField0_ & ~0x00002000);
          }
          result.secondaryAlternates_ = secondaryAlternates_;
        } else {
          result.secondaryAlternates_ = secondaryAlternatesBuilder_.build();
        }
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.getDefaultInstance()) return this;
        if (other.getRelativeStart() != 0) {
          setRelativeStart(other.getRelativeStart());
        }
        if (other.getRelativeEnd() != 0) {
          setRelativeEnd(other.getRelativeEnd());
        }
        if (!other.getReference().isEmpty()) {
          reference_ = other.reference_;
          onChanged();
        }
        if (!other.getAlternate().isEmpty()) {
          alternate_ = other.alternate_;
          onChanged();
        }
        if (other.getQuality() != 0F) {
          setQuality(other.getQuality());
        }
        if (other.type_ != 0) {
          setTypeValue(other.getTypeValue());
        }
        if (!other.getCall().isEmpty()) {
          call_ = other.call_;
          onChanged();
        }
        if (other.getFilterIndex() != 0) {
          setFilterIndex(other.getFilterIndex());
        }
        if (!other.idNonDefault_.isEmpty()) {
          if (idNonDefault_.isEmpty()) {
            idNonDefault_ = other.idNonDefault_;
            bitField0_ = (bitField0_ & ~0x00000100);
          } else {
            ensureIdNonDefaultIsMutable();
            idNonDefault_.addAll(other.idNonDefault_);
          }
          onChanged();
        }
        if (!other.infoKeyIndex_.isEmpty()) {
          if (infoKeyIndex_.isEmpty()) {
            infoKeyIndex_ = other.infoKeyIndex_;
            bitField0_ = (bitField0_ & ~0x00000200);
          } else {
            ensureInfoKeyIndexIsMutable();
            infoKeyIndex_.addAll(other.infoKeyIndex_);
          }
          onChanged();
        }
        if (!other.infoValue_.isEmpty()) {
          if (infoValue_.isEmpty()) {
            infoValue_ = other.infoValue_;
            bitField0_ = (bitField0_ & ~0x00000400);
          } else {
            ensureInfoValueIsMutable();
            infoValue_.addAll(other.infoValue_);
          }
          onChanged();
        }
        if (other.getFormatIndex() != 0) {
          setFormatIndex(other.getFormatIndex());
        }
        if (samplesBuilder_ == null) {
          if (!other.samples_.isEmpty()) {
            if (samples_.isEmpty()) {
              samples_ = other.samples_;
              bitField0_ = (bitField0_ & ~0x00001000);
            } else {
              ensureSamplesIsMutable();
              samples_.addAll(other.samples_);
            }
            onChanged();
          }
        } else {
          if (!other.samples_.isEmpty()) {
            if (samplesBuilder_.isEmpty()) {
              samplesBuilder_.dispose();
              samplesBuilder_ = null;
              samples_ = other.samples_;
              bitField0_ = (bitField0_ & ~0x00001000);
              samplesBuilder_ = 
                com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders ?
                   getSamplesFieldBuilder() : null;
            } else {
              samplesBuilder_.addAllMessages(other.samples_);
            }
          }
        }
        if (secondaryAlternatesBuilder_ == null) {
          if (!other.secondaryAlternates_.isEmpty()) {
            if (secondaryAlternates_.isEmpty()) {
              secondaryAlternates_ = other.secondaryAlternates_;
              bitField0_ = (bitField0_ & ~0x00002000);
            } else {
              ensureSecondaryAlternatesIsMutable();
              secondaryAlternates_.addAll(other.secondaryAlternates_);
            }
            onChanged();
          }
        } else {
          if (!other.secondaryAlternates_.isEmpty()) {
            if (secondaryAlternatesBuilder_.isEmpty()) {
              secondaryAlternatesBuilder_.dispose();
              secondaryAlternatesBuilder_ = null;
              secondaryAlternates_ = other.secondaryAlternates_;
              bitField0_ = (bitField0_ & ~0x00002000);
              secondaryAlternatesBuilder_ = 
                com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders ?
                   getSecondaryAlternatesFieldBuilder() : null;
            } else {
              secondaryAlternatesBuilder_.addAllMessages(other.secondaryAlternates_);
            }
          }
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private int relativeStart_ ;
      /**
       * <code>optional int32 relative_start = 1;</code>
       *
       * <pre>
       * 1 based
       * May contain negative values but it's not likely
       * </pre>
       */
      public int getRelativeStart() {
        return relativeStart_;
      }
      /**
       * <code>optional int32 relative_start = 1;</code>
       *
       * <pre>
       * 1 based
       * May contain negative values but it's not likely
       * </pre>
       */
      public Builder setRelativeStart(int value) {
        
        relativeStart_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 relative_start = 1;</code>
       *
       * <pre>
       * 1 based
       * May contain negative values but it's not likely
       * </pre>
       */
      public Builder clearRelativeStart() {
        
        relativeStart_ = 0;
        onChanged();
        return this;
      }

      private int relativeEnd_ ;
      /**
       * <code>optional int32 relative_end = 2;</code>
       *
       * <pre>
       * May contain negative values but it's not likely
       * </pre>
       */
      public int getRelativeEnd() {
        return relativeEnd_;
      }
      /**
       * <code>optional int32 relative_end = 2;</code>
       *
       * <pre>
       * May contain negative values but it's not likely
       * </pre>
       */
      public Builder setRelativeEnd(int value) {
        
        relativeEnd_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional int32 relative_end = 2;</code>
       *
       * <pre>
       * May contain negative values but it's not likely
       * </pre>
       */
      public Builder clearRelativeEnd() {
        
        relativeEnd_ = 0;
        onChanged();
        return this;
      }

      private java.lang.Object reference_ = "";
      /**
       * <code>optional string reference = 3;</code>
       */
      public java.lang.String getReference() {
        java.lang.Object ref = reference_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          reference_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string reference = 3;</code>
       */
      public com.google.protobuf.ByteString
          getReferenceBytes() {
        java.lang.Object ref = reference_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          reference_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string reference = 3;</code>
       */
      public Builder setReference(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        reference_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string reference = 3;</code>
       */
      public Builder clearReference() {
        
        reference_ = getDefaultInstance().getReference();
        onChanged();
        return this;
      }
      /**
       * <code>optional string reference = 3;</code>
       */
      public Builder setReferenceBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        reference_ = value;
        onChanged();
        return this;
      }

      private java.lang.Object alternate_ = "";
      /**
       * <code>optional string alternate = 4;</code>
       */
      public java.lang.String getAlternate() {
        java.lang.Object ref = alternate_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          alternate_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string alternate = 4;</code>
       */
      public com.google.protobuf.ByteString
          getAlternateBytes() {
        java.lang.Object ref = alternate_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          alternate_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string alternate = 4;</code>
       */
      public Builder setAlternate(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        alternate_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string alternate = 4;</code>
       */
      public Builder clearAlternate() {
        
        alternate_ = getDefaultInstance().getAlternate();
        onChanged();
        return this;
      }
      /**
       * <code>optional string alternate = 4;</code>
       */
      public Builder setAlternateBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        alternate_ = value;
        onChanged();
        return this;
      }

      private float quality_ ;
      /**
       * <code>optional float quality = 5;</code>
       */
      public float getQuality() {
        return quality_;
      }
      /**
       * <code>optional float quality = 5;</code>
       */
      public Builder setQuality(float value) {
        
        quality_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional float quality = 5;</code>
       */
      public Builder clearQuality() {
        
        quality_ = 0F;
        onChanged();
        return this;
      }

      private int type_ = 0;
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
       */
      public int getTypeValue() {
        return type_;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
       */
      public Builder setTypeValue(int value) {
        type_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType getType() {
        org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType result = org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.valueOf(type_);
        return result == null ? org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType.UNRECOGNIZED : result;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
       */
      public Builder setType(org.opencb.biodata.models.variant.protobuf.VariantProto.VariantType value) {
        if (value == null) {
          throw new NullPointerException();
        }
        
        type_ = value.getNumber();
        onChanged();
        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.VariantType type = 12;</code>
       */
      public Builder clearType() {
        
        type_ = 0;
        onChanged();
        return this;
      }

      private java.lang.Object call_ = "";
      /**
       * <code>optional string call = 13;</code>
       */
      public java.lang.String getCall() {
        java.lang.Object ref = call_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          call_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string call = 13;</code>
       */
      public com.google.protobuf.ByteString
          getCallBytes() {
        java.lang.Object ref = call_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          call_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string call = 13;</code>
       */
      public Builder setCall(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        call_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string call = 13;</code>
       */
      public Builder clearCall() {
        
        call_ = getDefaultInstance().getCall();
        onChanged();
        return this;
      }
      /**
       * <code>optional string call = 13;</code>
       */
      public Builder setCallBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        call_ = value;
        onChanged();
        return this;
      }

      private int filterIndex_ ;
      /**
       * <code>optional uint32 filter_index = 6;</code>
       */
      public int getFilterIndex() {
        return filterIndex_;
      }
      /**
       * <code>optional uint32 filter_index = 6;</code>
       */
      public Builder setFilterIndex(int value) {
        
        filterIndex_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional uint32 filter_index = 6;</code>
       */
      public Builder clearFilterIndex() {
        
        filterIndex_ = 0;
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList idNonDefault_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureIdNonDefaultIsMutable() {
        if (!((bitField0_ & 0x00000100) == 0x00000100)) {
          idNonDefault_ = new com.google.protobuf.LazyStringArrayList(idNonDefault_);
          bitField0_ |= 0x00000100;
         }
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getIdNonDefaultList() {
        return idNonDefault_.getUnmodifiableView();
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public int getIdNonDefaultCount() {
        return idNonDefault_.size();
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public java.lang.String getIdNonDefault(int index) {
        return idNonDefault_.get(index);
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public com.google.protobuf.ByteString
          getIdNonDefaultBytes(int index) {
        return idNonDefault_.getByteString(index);
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public Builder setIdNonDefault(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureIdNonDefaultIsMutable();
        idNonDefault_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public Builder addIdNonDefault(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureIdNonDefaultIsMutable();
        idNonDefault_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public Builder addAllIdNonDefault(
          java.lang.Iterable<java.lang.String> values) {
        ensureIdNonDefaultIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, idNonDefault_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public Builder clearIdNonDefault() {
        idNonDefault_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000100);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string id_non_default = 7;</code>
       */
      public Builder addIdNonDefaultBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureIdNonDefaultIsMutable();
        idNonDefault_.add(value);
        onChanged();
        return this;
      }

      private java.util.List<java.lang.Integer> infoKeyIndex_ = java.util.Collections.emptyList();
      private void ensureInfoKeyIndexIsMutable() {
        if (!((bitField0_ & 0x00000200) == 0x00000200)) {
          infoKeyIndex_ = new java.util.ArrayList<java.lang.Integer>(infoKeyIndex_);
          bitField0_ |= 0x00000200;
         }
      }
      /**
       * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
       */
      public java.util.List<java.lang.Integer>
          getInfoKeyIndexList() {
        return java.util.Collections.unmodifiableList(infoKeyIndex_);
      }
      /**
       * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
       */
      public int getInfoKeyIndexCount() {
        return infoKeyIndex_.size();
      }
      /**
       * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
       */
      public int getInfoKeyIndex(int index) {
        return infoKeyIndex_.get(index);
      }
      /**
       * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
       */
      public Builder setInfoKeyIndex(
          int index, int value) {
        ensureInfoKeyIndexIsMutable();
        infoKeyIndex_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
       */
      public Builder addInfoKeyIndex(int value) {
        ensureInfoKeyIndexIsMutable();
        infoKeyIndex_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
       */
      public Builder addAllInfoKeyIndex(
          java.lang.Iterable<? extends java.lang.Integer> values) {
        ensureInfoKeyIndexIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, infoKeyIndex_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated uint32 info_key_index = 8 [packed = true];</code>
       */
      public Builder clearInfoKeyIndex() {
        infoKeyIndex_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000200);
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList infoValue_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureInfoValueIsMutable() {
        if (!((bitField0_ & 0x00000400) == 0x00000400)) {
          infoValue_ = new com.google.protobuf.LazyStringArrayList(infoValue_);
          bitField0_ |= 0x00000400;
         }
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getInfoValueList() {
        return infoValue_.getUnmodifiableView();
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public int getInfoValueCount() {
        return infoValue_.size();
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public java.lang.String getInfoValue(int index) {
        return infoValue_.get(index);
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public com.google.protobuf.ByteString
          getInfoValueBytes(int index) {
        return infoValue_.getByteString(index);
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public Builder setInfoValue(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoValueIsMutable();
        infoValue_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public Builder addInfoValue(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoValueIsMutable();
        infoValue_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public Builder addAllInfoValue(
          java.lang.Iterable<java.lang.String> values) {
        ensureInfoValueIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, infoValue_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public Builder clearInfoValue() {
        infoValue_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000400);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_value = 9;</code>
       */
      public Builder addInfoValueBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureInfoValueIsMutable();
        infoValue_.add(value);
        onChanged();
        return this;
      }

      private int formatIndex_ ;
      /**
       * <code>optional uint32 formatIndex = 10;</code>
       */
      public int getFormatIndex() {
        return formatIndex_;
      }
      /**
       * <code>optional uint32 formatIndex = 10;</code>
       */
      public Builder setFormatIndex(int value) {
        
        formatIndex_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional uint32 formatIndex = 10;</code>
       */
      public Builder clearFormatIndex() {
        
        formatIndex_ = 0;
        onChanged();
        return this;
      }

      private java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample> samples_ =
        java.util.Collections.emptyList();
      private void ensureSamplesIsMutable() {
        if (!((bitField0_ & 0x00001000) == 0x00001000)) {
          samples_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample>(samples_);
          bitField0_ |= 0x00001000;
         }
      }

      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder> samplesBuilder_;

      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample> getSamplesList() {
        if (samplesBuilder_ == null) {
          return java.util.Collections.unmodifiableList(samples_);
        } else {
          return samplesBuilder_.getMessageList();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public int getSamplesCount() {
        if (samplesBuilder_ == null) {
          return samples_.size();
        } else {
          return samplesBuilder_.getCount();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample getSamples(int index) {
        if (samplesBuilder_ == null) {
          return samples_.get(index);
        } else {
          return samplesBuilder_.getMessage(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder setSamples(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample value) {
        if (samplesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSamplesIsMutable();
          samples_.set(index, value);
          onChanged();
        } else {
          samplesBuilder_.setMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder setSamples(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder builderForValue) {
        if (samplesBuilder_ == null) {
          ensureSamplesIsMutable();
          samples_.set(index, builderForValue.build());
          onChanged();
        } else {
          samplesBuilder_.setMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder addSamples(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample value) {
        if (samplesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSamplesIsMutable();
          samples_.add(value);
          onChanged();
        } else {
          samplesBuilder_.addMessage(value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder addSamples(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample value) {
        if (samplesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSamplesIsMutable();
          samples_.add(index, value);
          onChanged();
        } else {
          samplesBuilder_.addMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder addSamples(
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder builderForValue) {
        if (samplesBuilder_ == null) {
          ensureSamplesIsMutable();
          samples_.add(builderForValue.build());
          onChanged();
        } else {
          samplesBuilder_.addMessage(builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder addSamples(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder builderForValue) {
        if (samplesBuilder_ == null) {
          ensureSamplesIsMutable();
          samples_.add(index, builderForValue.build());
          onChanged();
        } else {
          samplesBuilder_.addMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder addAllSamples(
          java.lang.Iterable<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample> values) {
        if (samplesBuilder_ == null) {
          ensureSamplesIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, samples_);
          onChanged();
        } else {
          samplesBuilder_.addAllMessages(values);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder clearSamples() {
        if (samplesBuilder_ == null) {
          samples_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00001000);
          onChanged();
        } else {
          samplesBuilder_.clear();
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public Builder removeSamples(int index) {
        if (samplesBuilder_ == null) {
          ensureSamplesIsMutable();
          samples_.remove(index);
          onChanged();
        } else {
          samplesBuilder_.remove(index);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder getSamplesBuilder(
          int index) {
        return getSamplesFieldBuilder().getBuilder(index);
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder getSamplesOrBuilder(
          int index) {
        if (samplesBuilder_ == null) {
          return samples_.get(index);  } else {
          return samplesBuilder_.getMessageOrBuilder(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder> 
           getSamplesOrBuilderList() {
        if (samplesBuilder_ != null) {
          return samplesBuilder_.getMessageOrBuilderList();
        } else {
          return java.util.Collections.unmodifiableList(samples_);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder addSamplesBuilder() {
        return getSamplesFieldBuilder().addBuilder(
            org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder addSamplesBuilder(
          int index) {
        return getSamplesFieldBuilder().addBuilder(
            index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfSample samples = 11;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder> 
           getSamplesBuilderList() {
        return getSamplesFieldBuilder().getBuilderList();
      }
      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder> 
          getSamplesFieldBuilder() {
        if (samplesBuilder_ == null) {
          samplesBuilder_ = new com.google.protobuf.RepeatedFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSample.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSampleOrBuilder>(
                  samples_,
                  ((bitField0_ & 0x00001000) == 0x00001000),
                  getParentForChildren(),
                  isClean());
          samples_ = null;
        }
        return samplesBuilder_;
      }

      private java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate> secondaryAlternates_ =
        java.util.Collections.emptyList();
      private void ensureSecondaryAlternatesIsMutable() {
        if (!((bitField0_ & 0x00002000) == 0x00002000)) {
          secondaryAlternates_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate>(secondaryAlternates_);
          bitField0_ |= 0x00002000;
         }
      }

      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder> secondaryAlternatesBuilder_;

      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate> getSecondaryAlternatesList() {
        if (secondaryAlternatesBuilder_ == null) {
          return java.util.Collections.unmodifiableList(secondaryAlternates_);
        } else {
          return secondaryAlternatesBuilder_.getMessageList();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public int getSecondaryAlternatesCount() {
        if (secondaryAlternatesBuilder_ == null) {
          return secondaryAlternates_.size();
        } else {
          return secondaryAlternatesBuilder_.getCount();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate getSecondaryAlternates(int index) {
        if (secondaryAlternatesBuilder_ == null) {
          return secondaryAlternates_.get(index);
        } else {
          return secondaryAlternatesBuilder_.getMessage(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder setSecondaryAlternates(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate value) {
        if (secondaryAlternatesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSecondaryAlternatesIsMutable();
          secondaryAlternates_.set(index, value);
          onChanged();
        } else {
          secondaryAlternatesBuilder_.setMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder setSecondaryAlternates(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder builderForValue) {
        if (secondaryAlternatesBuilder_ == null) {
          ensureSecondaryAlternatesIsMutable();
          secondaryAlternates_.set(index, builderForValue.build());
          onChanged();
        } else {
          secondaryAlternatesBuilder_.setMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder addSecondaryAlternates(org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate value) {
        if (secondaryAlternatesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSecondaryAlternatesIsMutable();
          secondaryAlternates_.add(value);
          onChanged();
        } else {
          secondaryAlternatesBuilder_.addMessage(value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder addSecondaryAlternates(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate value) {
        if (secondaryAlternatesBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureSecondaryAlternatesIsMutable();
          secondaryAlternates_.add(index, value);
          onChanged();
        } else {
          secondaryAlternatesBuilder_.addMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder addSecondaryAlternates(
          org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder builderForValue) {
        if (secondaryAlternatesBuilder_ == null) {
          ensureSecondaryAlternatesIsMutable();
          secondaryAlternates_.add(builderForValue.build());
          onChanged();
        } else {
          secondaryAlternatesBuilder_.addMessage(builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder addSecondaryAlternates(
          int index, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder builderForValue) {
        if (secondaryAlternatesBuilder_ == null) {
          ensureSecondaryAlternatesIsMutable();
          secondaryAlternates_.add(index, builderForValue.build());
          onChanged();
        } else {
          secondaryAlternatesBuilder_.addMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder addAllSecondaryAlternates(
          java.lang.Iterable<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate> values) {
        if (secondaryAlternatesBuilder_ == null) {
          ensureSecondaryAlternatesIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, secondaryAlternates_);
          onChanged();
        } else {
          secondaryAlternatesBuilder_.addAllMessages(values);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder clearSecondaryAlternates() {
        if (secondaryAlternatesBuilder_ == null) {
          secondaryAlternates_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00002000);
          onChanged();
        } else {
          secondaryAlternatesBuilder_.clear();
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public Builder removeSecondaryAlternates(int index) {
        if (secondaryAlternatesBuilder_ == null) {
          ensureSecondaryAlternatesIsMutable();
          secondaryAlternates_.remove(index);
          onChanged();
        } else {
          secondaryAlternatesBuilder_.remove(index);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder getSecondaryAlternatesBuilder(
          int index) {
        return getSecondaryAlternatesFieldBuilder().getBuilder(index);
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder getSecondaryAlternatesOrBuilder(
          int index) {
        if (secondaryAlternatesBuilder_ == null) {
          return secondaryAlternates_.get(index);  } else {
          return secondaryAlternatesBuilder_.getMessageOrBuilder(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder> 
           getSecondaryAlternatesOrBuilderList() {
        if (secondaryAlternatesBuilder_ != null) {
          return secondaryAlternatesBuilder_.getMessageOrBuilderList();
        } else {
          return java.util.Collections.unmodifiableList(secondaryAlternates_);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder addSecondaryAlternatesBuilder() {
        return getSecondaryAlternatesFieldBuilder().addBuilder(
            org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder addSecondaryAlternatesBuilder(
          int index) {
        return getSecondaryAlternatesFieldBuilder().addBuilder(
            index, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.AlternateCoordinate secondaryAlternates = 14;</code>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder> 
           getSecondaryAlternatesBuilderList() {
        return getSecondaryAlternatesFieldBuilder().getBuilderList();
      }
      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder> 
          getSecondaryAlternatesFieldBuilder() {
        if (secondaryAlternatesBuilder_ == null) {
          secondaryAlternatesBuilder_ = new com.google.protobuf.RepeatedFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinate.Builder, org.opencb.biodata.models.variant.protobuf.VariantProto.AlternateCoordinateOrBuilder>(
                  secondaryAlternates_,
                  ((bitField0_ & 0x00002000) == 0x00002000),
                  getParentForChildren(),
                  isClean());
          secondaryAlternates_ = null;
        }
        return secondaryAlternatesBuilder_;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VcfRecord)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VcfRecord)
    private static final org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord();
    }

    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<VcfRecord>
        PARSER = new com.google.protobuf.AbstractParser<VcfRecord>() {
      public VcfRecord parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new VcfRecord(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<VcfRecord> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<VcfRecord> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface FieldsOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.Fields)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>repeated string info_keys = 1;</code>
     */
    com.google.protobuf.ProtocolStringList
        getInfoKeysList();
    /**
     * <code>repeated string info_keys = 1;</code>
     */
    int getInfoKeysCount();
    /**
     * <code>repeated string info_keys = 1;</code>
     */
    java.lang.String getInfoKeys(int index);
    /**
     * <code>repeated string info_keys = 1;</code>
     */
    com.google.protobuf.ByteString
        getInfoKeysBytes(int index);

    /**
     * <code>repeated uint32 default_info_keys = 2;</code>
     */
    java.util.List<java.lang.Integer> getDefaultInfoKeysList();
    /**
     * <code>repeated uint32 default_info_keys = 2;</code>
     */
    int getDefaultInfoKeysCount();
    /**
     * <code>repeated uint32 default_info_keys = 2;</code>
     */
    int getDefaultInfoKeys(int index);

    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    com.google.protobuf.ProtocolStringList
        getFiltersList();
    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    int getFiltersCount();
    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    java.lang.String getFilters(int index);
    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    com.google.protobuf.ByteString
        getFiltersBytes(int index);

    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    com.google.protobuf.ProtocolStringList
        getFormatsList();
    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    int getFormatsCount();
    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    java.lang.String getFormats(int index);
    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    com.google.protobuf.ByteString
        getFormatsBytes(int index);

    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    com.google.protobuf.ProtocolStringList
        getGtsList();
    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    int getGtsCount();
    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    java.lang.String getGts(int index);
    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    com.google.protobuf.ByteString
        getGtsBytes(int index);
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Fields}
   */
  public  static final class Fields extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.Fields)
      FieldsOrBuilder {
    // Use Fields.newBuilder() to construct.
    private Fields(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private Fields() {
      infoKeys_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      defaultInfoKeys_ = java.util.Collections.emptyList();
      filters_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      formats_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      gts_ = com.google.protobuf.LazyStringArrayList.EMPTY;
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private Fields(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
                infoKeys_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000001;
              }
              infoKeys_.add(s);
              break;
            }
            case 16: {
              if (!((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
                defaultInfoKeys_ = new java.util.ArrayList<java.lang.Integer>();
                mutable_bitField0_ |= 0x00000002;
              }
              defaultInfoKeys_.add(input.readUInt32());
              break;
            }
            case 18: {
              int length = input.readRawVarint32();
              int limit = input.pushLimit(length);
              if (!((mutable_bitField0_ & 0x00000002) == 0x00000002) && input.getBytesUntilLimit() > 0) {
                defaultInfoKeys_ = new java.util.ArrayList<java.lang.Integer>();
                mutable_bitField0_ |= 0x00000002;
              }
              while (input.getBytesUntilLimit() > 0) {
                defaultInfoKeys_.add(input.readUInt32());
              }
              input.popLimit(limit);
              break;
            }
            case 26: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
                filters_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000004;
              }
              filters_.add(s);
              break;
            }
            case 42: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000008) == 0x00000008)) {
                formats_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000008;
              }
              formats_.add(s);
              break;
            }
            case 50: {
              String s = input.readStringRequireUtf8();
              if (!((mutable_bitField0_ & 0x00000010) == 0x00000010)) {
                gts_ = new com.google.protobuf.LazyStringArrayList();
                mutable_bitField0_ |= 0x00000010;
              }
              gts_.add(s);
              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000001) == 0x00000001)) {
          infoKeys_ = infoKeys_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00000002) == 0x00000002)) {
          defaultInfoKeys_ = java.util.Collections.unmodifiableList(defaultInfoKeys_);
        }
        if (((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
          filters_ = filters_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00000008) == 0x00000008)) {
          formats_ = formats_.getUnmodifiableView();
        }
        if (((mutable_bitField0_ & 0x00000010) == 0x00000010)) {
          gts_ = gts_.getUnmodifiableView();
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_Fields_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_Fields_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder.class);
    }

    public static final int INFO_KEYS_FIELD_NUMBER = 1;
    private com.google.protobuf.LazyStringList infoKeys_;
    /**
     * <code>repeated string info_keys = 1;</code>
     */
    public com.google.protobuf.ProtocolStringList
        getInfoKeysList() {
      return infoKeys_;
    }
    /**
     * <code>repeated string info_keys = 1;</code>
     */
    public int getInfoKeysCount() {
      return infoKeys_.size();
    }
    /**
     * <code>repeated string info_keys = 1;</code>
     */
    public java.lang.String getInfoKeys(int index) {
      return infoKeys_.get(index);
    }
    /**
     * <code>repeated string info_keys = 1;</code>
     */
    public com.google.protobuf.ByteString
        getInfoKeysBytes(int index) {
      return infoKeys_.getByteString(index);
    }

    public static final int DEFAULT_INFO_KEYS_FIELD_NUMBER = 2;
    private java.util.List<java.lang.Integer> defaultInfoKeys_;
    /**
     * <code>repeated uint32 default_info_keys = 2;</code>
     */
    public java.util.List<java.lang.Integer>
        getDefaultInfoKeysList() {
      return defaultInfoKeys_;
    }
    /**
     * <code>repeated uint32 default_info_keys = 2;</code>
     */
    public int getDefaultInfoKeysCount() {
      return defaultInfoKeys_.size();
    }
    /**
     * <code>repeated uint32 default_info_keys = 2;</code>
     */
    public int getDefaultInfoKeys(int index) {
      return defaultInfoKeys_.get(index);
    }
    private int defaultInfoKeysMemoizedSerializedSize = -1;

    public static final int FILTERS_FIELD_NUMBER = 3;
    private com.google.protobuf.LazyStringList filters_;
    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    public com.google.protobuf.ProtocolStringList
        getFiltersList() {
      return filters_;
    }
    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    public int getFiltersCount() {
      return filters_.size();
    }
    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    public java.lang.String getFilters(int index) {
      return filters_.get(index);
    }
    /**
     * <code>repeated string filters = 3;</code>
     *
     * <pre>
     * Possible filter compositions. Delimited by ';'
     * Where the first filter is the default one
     * </pre>
     */
    public com.google.protobuf.ByteString
        getFiltersBytes(int index) {
      return filters_.getByteString(index);
    }

    public static final int FORMATS_FIELD_NUMBER = 5;
    private com.google.protobuf.LazyStringList formats_;
    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    public com.google.protobuf.ProtocolStringList
        getFormatsList() {
      return formats_;
    }
    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    public int getFormatsCount() {
      return formats_.size();
    }
    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    public java.lang.String getFormats(int index) {
      return formats_.get(index);
    }
    /**
     * <code>repeated string formats = 5;</code>
     *
     * <pre>
     * Possible formats compositions. Delimited by ':'
     * Where the first format is the default one
     * </pre>
     */
    public com.google.protobuf.ByteString
        getFormatsBytes(int index) {
      return formats_.getByteString(index);
    }

    public static final int GTS_FIELD_NUMBER = 6;
    private com.google.protobuf.LazyStringList gts_;
    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    public com.google.protobuf.ProtocolStringList
        getGtsList() {
      return gts_;
    }
    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    public int getGtsCount() {
      return gts_.size();
    }
    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    public java.lang.String getGts(int index) {
      return gts_.get(index);
    }
    /**
     * <code>repeated string gts = 6;</code>
     *
     * <pre>
     * Possible genotypes seen on the slice.
     * Where the first GT is the default one
     * </pre>
     */
    public com.google.protobuf.ByteString
        getGtsBytes(int index) {
      return gts_.getByteString(index);
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      getSerializedSize();
      for (int i = 0; i < infoKeys_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, infoKeys_.getRaw(i));
      }
      if (getDefaultInfoKeysList().size() > 0) {
        output.writeRawVarint32(18);
        output.writeRawVarint32(defaultInfoKeysMemoizedSerializedSize);
      }
      for (int i = 0; i < defaultInfoKeys_.size(); i++) {
        output.writeUInt32NoTag(defaultInfoKeys_.get(i));
      }
      for (int i = 0; i < filters_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 3, filters_.getRaw(i));
      }
      for (int i = 0; i < formats_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 5, formats_.getRaw(i));
      }
      for (int i = 0; i < gts_.size(); i++) {
        com.google.protobuf.GeneratedMessage.writeString(output, 6, gts_.getRaw(i));
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      {
        int dataSize = 0;
        for (int i = 0; i < infoKeys_.size(); i++) {
          dataSize += computeStringSizeNoTag(infoKeys_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getInfoKeysList().size();
      }
      {
        int dataSize = 0;
        for (int i = 0; i < defaultInfoKeys_.size(); i++) {
          dataSize += com.google.protobuf.CodedOutputStream
            .computeUInt32SizeNoTag(defaultInfoKeys_.get(i));
        }
        size += dataSize;
        if (!getDefaultInfoKeysList().isEmpty()) {
          size += 1;
          size += com.google.protobuf.CodedOutputStream
              .computeInt32SizeNoTag(dataSize);
        }
        defaultInfoKeysMemoizedSerializedSize = dataSize;
      }
      {
        int dataSize = 0;
        for (int i = 0; i < filters_.size(); i++) {
          dataSize += computeStringSizeNoTag(filters_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getFiltersList().size();
      }
      {
        int dataSize = 0;
        for (int i = 0; i < formats_.size(); i++) {
          dataSize += computeStringSizeNoTag(formats_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getFormatsList().size();
      }
      {
        int dataSize = 0;
        for (int i = 0; i < gts_.size(); i++) {
          dataSize += computeStringSizeNoTag(gts_.getRaw(i));
        }
        size += dataSize;
        size += 1 * getGtsList().size();
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.Fields}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.Fields)
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.FieldsOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_Fields_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_Fields_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
        }
      }
      public Builder clear() {
        super.clear();
        infoKeys_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000001);
        defaultInfoKeys_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000002);
        filters_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000004);
        formats_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000008);
        gts_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000010);
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_Fields_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields build() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields result = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields(this);
        int from_bitField0_ = bitField0_;
        if (((bitField0_ & 0x00000001) == 0x00000001)) {
          infoKeys_ = infoKeys_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000001);
        }
        result.infoKeys_ = infoKeys_;
        if (((bitField0_ & 0x00000002) == 0x00000002)) {
          defaultInfoKeys_ = java.util.Collections.unmodifiableList(defaultInfoKeys_);
          bitField0_ = (bitField0_ & ~0x00000002);
        }
        result.defaultInfoKeys_ = defaultInfoKeys_;
        if (((bitField0_ & 0x00000004) == 0x00000004)) {
          filters_ = filters_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000004);
        }
        result.filters_ = filters_;
        if (((bitField0_ & 0x00000008) == 0x00000008)) {
          formats_ = formats_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000008);
        }
        result.formats_ = formats_;
        if (((bitField0_ & 0x00000010) == 0x00000010)) {
          gts_ = gts_.getUnmodifiableView();
          bitField0_ = (bitField0_ & ~0x00000010);
        }
        result.gts_ = gts_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.getDefaultInstance()) return this;
        if (!other.infoKeys_.isEmpty()) {
          if (infoKeys_.isEmpty()) {
            infoKeys_ = other.infoKeys_;
            bitField0_ = (bitField0_ & ~0x00000001);
          } else {
            ensureInfoKeysIsMutable();
            infoKeys_.addAll(other.infoKeys_);
          }
          onChanged();
        }
        if (!other.defaultInfoKeys_.isEmpty()) {
          if (defaultInfoKeys_.isEmpty()) {
            defaultInfoKeys_ = other.defaultInfoKeys_;
            bitField0_ = (bitField0_ & ~0x00000002);
          } else {
            ensureDefaultInfoKeysIsMutable();
            defaultInfoKeys_.addAll(other.defaultInfoKeys_);
          }
          onChanged();
        }
        if (!other.filters_.isEmpty()) {
          if (filters_.isEmpty()) {
            filters_ = other.filters_;
            bitField0_ = (bitField0_ & ~0x00000004);
          } else {
            ensureFiltersIsMutable();
            filters_.addAll(other.filters_);
          }
          onChanged();
        }
        if (!other.formats_.isEmpty()) {
          if (formats_.isEmpty()) {
            formats_ = other.formats_;
            bitField0_ = (bitField0_ & ~0x00000008);
          } else {
            ensureFormatsIsMutable();
            formats_.addAll(other.formats_);
          }
          onChanged();
        }
        if (!other.gts_.isEmpty()) {
          if (gts_.isEmpty()) {
            gts_ = other.gts_;
            bitField0_ = (bitField0_ & ~0x00000010);
          } else {
            ensureGtsIsMutable();
            gts_.addAll(other.gts_);
          }
          onChanged();
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private com.google.protobuf.LazyStringList infoKeys_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureInfoKeysIsMutable() {
        if (!((bitField0_ & 0x00000001) == 0x00000001)) {
          infoKeys_ = new com.google.protobuf.LazyStringArrayList(infoKeys_);
          bitField0_ |= 0x00000001;
         }
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public com.google.protobuf.ProtocolStringList
          getInfoKeysList() {
        return infoKeys_.getUnmodifiableView();
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public int getInfoKeysCount() {
        return infoKeys_.size();
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public java.lang.String getInfoKeys(int index) {
        return infoKeys_.get(index);
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public com.google.protobuf.ByteString
          getInfoKeysBytes(int index) {
        return infoKeys_.getByteString(index);
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public Builder setInfoKeys(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoKeysIsMutable();
        infoKeys_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public Builder addInfoKeys(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureInfoKeysIsMutable();
        infoKeys_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public Builder addAllInfoKeys(
          java.lang.Iterable<java.lang.String> values) {
        ensureInfoKeysIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, infoKeys_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public Builder clearInfoKeys() {
        infoKeys_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000001);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string info_keys = 1;</code>
       */
      public Builder addInfoKeysBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureInfoKeysIsMutable();
        infoKeys_.add(value);
        onChanged();
        return this;
      }

      private java.util.List<java.lang.Integer> defaultInfoKeys_ = java.util.Collections.emptyList();
      private void ensureDefaultInfoKeysIsMutable() {
        if (!((bitField0_ & 0x00000002) == 0x00000002)) {
          defaultInfoKeys_ = new java.util.ArrayList<java.lang.Integer>(defaultInfoKeys_);
          bitField0_ |= 0x00000002;
         }
      }
      /**
       * <code>repeated uint32 default_info_keys = 2;</code>
       */
      public java.util.List<java.lang.Integer>
          getDefaultInfoKeysList() {
        return java.util.Collections.unmodifiableList(defaultInfoKeys_);
      }
      /**
       * <code>repeated uint32 default_info_keys = 2;</code>
       */
      public int getDefaultInfoKeysCount() {
        return defaultInfoKeys_.size();
      }
      /**
       * <code>repeated uint32 default_info_keys = 2;</code>
       */
      public int getDefaultInfoKeys(int index) {
        return defaultInfoKeys_.get(index);
      }
      /**
       * <code>repeated uint32 default_info_keys = 2;</code>
       */
      public Builder setDefaultInfoKeys(
          int index, int value) {
        ensureDefaultInfoKeysIsMutable();
        defaultInfoKeys_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated uint32 default_info_keys = 2;</code>
       */
      public Builder addDefaultInfoKeys(int value) {
        ensureDefaultInfoKeysIsMutable();
        defaultInfoKeys_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated uint32 default_info_keys = 2;</code>
       */
      public Builder addAllDefaultInfoKeys(
          java.lang.Iterable<? extends java.lang.Integer> values) {
        ensureDefaultInfoKeysIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, defaultInfoKeys_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated uint32 default_info_keys = 2;</code>
       */
      public Builder clearDefaultInfoKeys() {
        defaultInfoKeys_ = java.util.Collections.emptyList();
        bitField0_ = (bitField0_ & ~0x00000002);
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList filters_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureFiltersIsMutable() {
        if (!((bitField0_ & 0x00000004) == 0x00000004)) {
          filters_ = new com.google.protobuf.LazyStringArrayList(filters_);
          bitField0_ |= 0x00000004;
         }
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public com.google.protobuf.ProtocolStringList
          getFiltersList() {
        return filters_.getUnmodifiableView();
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public int getFiltersCount() {
        return filters_.size();
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public java.lang.String getFilters(int index) {
        return filters_.get(index);
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public com.google.protobuf.ByteString
          getFiltersBytes(int index) {
        return filters_.getByteString(index);
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public Builder setFilters(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureFiltersIsMutable();
        filters_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public Builder addFilters(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureFiltersIsMutable();
        filters_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public Builder addAllFilters(
          java.lang.Iterable<java.lang.String> values) {
        ensureFiltersIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, filters_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public Builder clearFilters() {
        filters_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000004);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string filters = 3;</code>
       *
       * <pre>
       * Possible filter compositions. Delimited by ';'
       * Where the first filter is the default one
       * </pre>
       */
      public Builder addFiltersBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureFiltersIsMutable();
        filters_.add(value);
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList formats_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureFormatsIsMutable() {
        if (!((bitField0_ & 0x00000008) == 0x00000008)) {
          formats_ = new com.google.protobuf.LazyStringArrayList(formats_);
          bitField0_ |= 0x00000008;
         }
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public com.google.protobuf.ProtocolStringList
          getFormatsList() {
        return formats_.getUnmodifiableView();
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public int getFormatsCount() {
        return formats_.size();
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public java.lang.String getFormats(int index) {
        return formats_.get(index);
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public com.google.protobuf.ByteString
          getFormatsBytes(int index) {
        return formats_.getByteString(index);
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public Builder setFormats(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureFormatsIsMutable();
        formats_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public Builder addFormats(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureFormatsIsMutable();
        formats_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public Builder addAllFormats(
          java.lang.Iterable<java.lang.String> values) {
        ensureFormatsIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, formats_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public Builder clearFormats() {
        formats_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000008);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string formats = 5;</code>
       *
       * <pre>
       * Possible formats compositions. Delimited by ':'
       * Where the first format is the default one
       * </pre>
       */
      public Builder addFormatsBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureFormatsIsMutable();
        formats_.add(value);
        onChanged();
        return this;
      }

      private com.google.protobuf.LazyStringList gts_ = com.google.protobuf.LazyStringArrayList.EMPTY;
      private void ensureGtsIsMutable() {
        if (!((bitField0_ & 0x00000010) == 0x00000010)) {
          gts_ = new com.google.protobuf.LazyStringArrayList(gts_);
          bitField0_ |= 0x00000010;
         }
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public com.google.protobuf.ProtocolStringList
          getGtsList() {
        return gts_.getUnmodifiableView();
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public int getGtsCount() {
        return gts_.size();
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public java.lang.String getGts(int index) {
        return gts_.get(index);
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public com.google.protobuf.ByteString
          getGtsBytes(int index) {
        return gts_.getByteString(index);
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public Builder setGts(
          int index, java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureGtsIsMutable();
        gts_.set(index, value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public Builder addGts(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  ensureGtsIsMutable();
        gts_.add(value);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public Builder addAllGts(
          java.lang.Iterable<java.lang.String> values) {
        ensureGtsIsMutable();
        com.google.protobuf.AbstractMessageLite.Builder.addAll(
            values, gts_);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public Builder clearGts() {
        gts_ = com.google.protobuf.LazyStringArrayList.EMPTY;
        bitField0_ = (bitField0_ & ~0x00000010);
        onChanged();
        return this;
      }
      /**
       * <code>repeated string gts = 6;</code>
       *
       * <pre>
       * Possible genotypes seen on the slice.
       * Where the first GT is the default one
       * </pre>
       */
      public Builder addGtsBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        ensureGtsIsMutable();
        gts_.add(value);
        onChanged();
        return this;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.Fields)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.Fields)
    private static final org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields();
    }

    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<Fields>
        PARSER = new com.google.protobuf.AbstractParser<Fields>() {
      public Fields parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new Fields(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<Fields> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<Fields> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  public interface VcfSliceOrBuilder extends
      // @@protoc_insertion_point(interface_extends:org.opencb.biodata.models.variant.protobuf.VcfSlice)
      com.google.protobuf.MessageOrBuilder {

    /**
     * <code>optional string chromosome = 1;</code>
     */
    java.lang.String getChromosome();
    /**
     * <code>optional string chromosome = 1;</code>
     */
    com.google.protobuf.ByteString
        getChromosomeBytes();

    /**
     * <code>optional uint32 position = 2;</code>
     */
    int getPosition();

    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord> 
        getRecordsList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord getRecords(int index);
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    int getRecordsCount();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder> 
        getRecordsOrBuilderList();
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder getRecordsOrBuilder(
        int index);

    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
     */
    boolean hasFields();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields getFields();
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
     */
    org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.FieldsOrBuilder getFieldsOrBuilder();
  }
  /**
   * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VcfSlice}
   */
  public  static final class VcfSlice extends
      com.google.protobuf.GeneratedMessage implements
      // @@protoc_insertion_point(message_implements:org.opencb.biodata.models.variant.protobuf.VcfSlice)
      VcfSliceOrBuilder {
    // Use VcfSlice.newBuilder() to construct.
    private VcfSlice(com.google.protobuf.GeneratedMessage.Builder<?> builder) {
      super(builder);
    }
    private VcfSlice() {
      chromosome_ = "";
      position_ = 0;
      records_ = java.util.Collections.emptyList();
    }

    @java.lang.Override
    public final com.google.protobuf.UnknownFieldSet
    getUnknownFields() {
      return com.google.protobuf.UnknownFieldSet.getDefaultInstance();
    }
    private VcfSlice(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry) {
      this();
      int mutable_bitField0_ = 0;
      try {
        boolean done = false;
        while (!done) {
          int tag = input.readTag();
          switch (tag) {
            case 0:
              done = true;
              break;
            default: {
              if (!input.skipField(tag)) {
                done = true;
              }
              break;
            }
            case 10: {
              String s = input.readStringRequireUtf8();

              chromosome_ = s;
              break;
            }
            case 16: {

              position_ = input.readUInt32();
              break;
            }
            case 26: {
              if (!((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
                records_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord>();
                mutable_bitField0_ |= 0x00000004;
              }
              records_.add(input.readMessage(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.parser(), extensionRegistry));
              break;
            }
            case 34: {
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder subBuilder = null;
              if (fields_ != null) {
                subBuilder = fields_.toBuilder();
              }
              fields_ = input.readMessage(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.parser(), extensionRegistry);
              if (subBuilder != null) {
                subBuilder.mergeFrom(fields_);
                fields_ = subBuilder.buildPartial();
              }

              break;
            }
          }
        }
      } catch (com.google.protobuf.InvalidProtocolBufferException e) {
        throw new RuntimeException(e.setUnfinishedMessage(this));
      } catch (java.io.IOException e) {
        throw new RuntimeException(
            new com.google.protobuf.InvalidProtocolBufferException(
                e.getMessage()).setUnfinishedMessage(this));
      } finally {
        if (((mutable_bitField0_ & 0x00000004) == 0x00000004)) {
          records_ = java.util.Collections.unmodifiableList(records_);
        }
        makeExtensionsImmutable();
      }
    }
    public static final com.google.protobuf.Descriptors.Descriptor
        getDescriptor() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_descriptor;
    }

    protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
        internalGetFieldAccessorTable() {
      return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_fieldAccessorTable
          .ensureFieldAccessorsInitialized(
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice.Builder.class);
    }

    private int bitField0_;
    public static final int CHROMOSOME_FIELD_NUMBER = 1;
    private volatile java.lang.Object chromosome_;
    /**
     * <code>optional string chromosome = 1;</code>
     */
    public java.lang.String getChromosome() {
      java.lang.Object ref = chromosome_;
      if (ref instanceof java.lang.String) {
        return (java.lang.String) ref;
      } else {
        com.google.protobuf.ByteString bs = 
            (com.google.protobuf.ByteString) ref;
        java.lang.String s = bs.toStringUtf8();
        chromosome_ = s;
        return s;
      }
    }
    /**
     * <code>optional string chromosome = 1;</code>
     */
    public com.google.protobuf.ByteString
        getChromosomeBytes() {
      java.lang.Object ref = chromosome_;
      if (ref instanceof java.lang.String) {
        com.google.protobuf.ByteString b = 
            com.google.protobuf.ByteString.copyFromUtf8(
                (java.lang.String) ref);
        chromosome_ = b;
        return b;
      } else {
        return (com.google.protobuf.ByteString) ref;
      }
    }

    public static final int POSITION_FIELD_NUMBER = 2;
    private int position_;
    /**
     * <code>optional uint32 position = 2;</code>
     */
    public int getPosition() {
      return position_;
    }

    public static final int RECORDS_FIELD_NUMBER = 3;
    private java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord> records_;
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    public java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord> getRecordsList() {
      return records_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder> 
        getRecordsOrBuilderList() {
      return records_;
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    public int getRecordsCount() {
      return records_.size();
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord getRecords(int index) {
      return records_.get(index);
    }
    /**
     * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
     *
     * <pre>
     * List of records (lines)
     * </pre>
     */
    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder getRecordsOrBuilder(
        int index) {
      return records_.get(index);
    }

    public static final int FIELDS_FIELD_NUMBER = 4;
    private org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields fields_;
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
     */
    public boolean hasFields() {
      return fields_ != null;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields getFields() {
      return fields_ == null ? org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.getDefaultInstance() : fields_;
    }
    /**
     * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
     */
    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.FieldsOrBuilder getFieldsOrBuilder() {
      return getFields();
    }

    private byte memoizedIsInitialized = -1;
    public final boolean isInitialized() {
      byte isInitialized = memoizedIsInitialized;
      if (isInitialized == 1) return true;
      if (isInitialized == 0) return false;

      memoizedIsInitialized = 1;
      return true;
    }

    public void writeTo(com.google.protobuf.CodedOutputStream output)
                        throws java.io.IOException {
      if (!getChromosomeBytes().isEmpty()) {
        com.google.protobuf.GeneratedMessage.writeString(output, 1, chromosome_);
      }
      if (position_ != 0) {
        output.writeUInt32(2, position_);
      }
      for (int i = 0; i < records_.size(); i++) {
        output.writeMessage(3, records_.get(i));
      }
      if (fields_ != null) {
        output.writeMessage(4, getFields());
      }
    }

    public int getSerializedSize() {
      int size = memoizedSize;
      if (size != -1) return size;

      size = 0;
      if (!getChromosomeBytes().isEmpty()) {
        size += com.google.protobuf.GeneratedMessage.computeStringSize(1, chromosome_);
      }
      if (position_ != 0) {
        size += com.google.protobuf.CodedOutputStream
          .computeUInt32Size(2, position_);
      }
      for (int i = 0; i < records_.size(); i++) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(3, records_.get(i));
      }
      if (fields_ != null) {
        size += com.google.protobuf.CodedOutputStream
          .computeMessageSize(4, getFields());
      }
      memoizedSize = size;
      return size;
    }

    private static final long serialVersionUID = 0L;
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(
        com.google.protobuf.ByteString data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(
        com.google.protobuf.ByteString data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(byte[] data)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(
        byte[] data,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws com.google.protobuf.InvalidProtocolBufferException {
      return PARSER.parseFrom(data, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseDelimitedFrom(java.io.InputStream input)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseDelimitedFrom(
        java.io.InputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseDelimitedFrom(input, extensionRegistry);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(
        com.google.protobuf.CodedInputStream input)
        throws java.io.IOException {
      return PARSER.parseFrom(input);
    }
    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parseFrom(
        com.google.protobuf.CodedInputStream input,
        com.google.protobuf.ExtensionRegistryLite extensionRegistry)
        throws java.io.IOException {
      return PARSER.parseFrom(input, extensionRegistry);
    }

    public Builder newBuilderForType() { return newBuilder(); }
    public static Builder newBuilder() {
      return DEFAULT_INSTANCE.toBuilder();
    }
    public static Builder newBuilder(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice prototype) {
      return DEFAULT_INSTANCE.toBuilder().mergeFrom(prototype);
    }
    public Builder toBuilder() {
      return this == DEFAULT_INSTANCE
          ? new Builder() : new Builder().mergeFrom(this);
    }

    @java.lang.Override
    protected Builder newBuilderForType(
        com.google.protobuf.GeneratedMessage.BuilderParent parent) {
      Builder builder = new Builder(parent);
      return builder;
    }
    /**
     * Protobuf type {@code org.opencb.biodata.models.variant.protobuf.VcfSlice}
     */
    public static final class Builder extends
        com.google.protobuf.GeneratedMessage.Builder<Builder> implements
        // @@protoc_insertion_point(builder_implements:org.opencb.biodata.models.variant.protobuf.VcfSlice)
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSliceOrBuilder {
      public static final com.google.protobuf.Descriptors.Descriptor
          getDescriptor() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_descriptor;
      }

      protected com.google.protobuf.GeneratedMessage.FieldAccessorTable
          internalGetFieldAccessorTable() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_fieldAccessorTable
            .ensureFieldAccessorsInitialized(
                org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice.class, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice.Builder.class);
      }

      // Construct using org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice.newBuilder()
      private Builder() {
        maybeForceBuilderInitialization();
      }

      private Builder(
          com.google.protobuf.GeneratedMessage.BuilderParent parent) {
        super(parent);
        maybeForceBuilderInitialization();
      }
      private void maybeForceBuilderInitialization() {
        if (com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders) {
          getRecordsFieldBuilder();
        }
      }
      public Builder clear() {
        super.clear();
        chromosome_ = "";

        position_ = 0;

        if (recordsBuilder_ == null) {
          records_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000004);
        } else {
          recordsBuilder_.clear();
        }
        if (fieldsBuilder_ == null) {
          fields_ = null;
        } else {
          fields_ = null;
          fieldsBuilder_ = null;
        }
        return this;
      }

      public com.google.protobuf.Descriptors.Descriptor
          getDescriptorForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_descriptor;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice getDefaultInstanceForType() {
        return org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice.getDefaultInstance();
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice build() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice result = buildPartial();
        if (!result.isInitialized()) {
          throw newUninitializedMessageException(result);
        }
        return result;
      }

      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice buildPartial() {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice result = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice(this);
        int from_bitField0_ = bitField0_;
        int to_bitField0_ = 0;
        result.chromosome_ = chromosome_;
        result.position_ = position_;
        if (recordsBuilder_ == null) {
          if (((bitField0_ & 0x00000004) == 0x00000004)) {
            records_ = java.util.Collections.unmodifiableList(records_);
            bitField0_ = (bitField0_ & ~0x00000004);
          }
          result.records_ = records_;
        } else {
          result.records_ = recordsBuilder_.build();
        }
        if (fieldsBuilder_ == null) {
          result.fields_ = fields_;
        } else {
          result.fields_ = fieldsBuilder_.build();
        }
        result.bitField0_ = to_bitField0_;
        onBuilt();
        return result;
      }

      public Builder mergeFrom(com.google.protobuf.Message other) {
        if (other instanceof org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice) {
          return mergeFrom((org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice)other);
        } else {
          super.mergeFrom(other);
          return this;
        }
      }

      public Builder mergeFrom(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice other) {
        if (other == org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice.getDefaultInstance()) return this;
        if (!other.getChromosome().isEmpty()) {
          chromosome_ = other.chromosome_;
          onChanged();
        }
        if (other.getPosition() != 0) {
          setPosition(other.getPosition());
        }
        if (recordsBuilder_ == null) {
          if (!other.records_.isEmpty()) {
            if (records_.isEmpty()) {
              records_ = other.records_;
              bitField0_ = (bitField0_ & ~0x00000004);
            } else {
              ensureRecordsIsMutable();
              records_.addAll(other.records_);
            }
            onChanged();
          }
        } else {
          if (!other.records_.isEmpty()) {
            if (recordsBuilder_.isEmpty()) {
              recordsBuilder_.dispose();
              recordsBuilder_ = null;
              records_ = other.records_;
              bitField0_ = (bitField0_ & ~0x00000004);
              recordsBuilder_ = 
                com.google.protobuf.GeneratedMessage.alwaysUseFieldBuilders ?
                   getRecordsFieldBuilder() : null;
            } else {
              recordsBuilder_.addAllMessages(other.records_);
            }
          }
        }
        if (other.hasFields()) {
          mergeFields(other.getFields());
        }
        onChanged();
        return this;
      }

      public final boolean isInitialized() {
        return true;
      }

      public Builder mergeFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws java.io.IOException {
        org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice parsedMessage = null;
        try {
          parsedMessage = PARSER.parsePartialFrom(input, extensionRegistry);
        } catch (com.google.protobuf.InvalidProtocolBufferException e) {
          parsedMessage = (org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice) e.getUnfinishedMessage();
          throw e;
        } finally {
          if (parsedMessage != null) {
            mergeFrom(parsedMessage);
          }
        }
        return this;
      }
      private int bitField0_;

      private java.lang.Object chromosome_ = "";
      /**
       * <code>optional string chromosome = 1;</code>
       */
      public java.lang.String getChromosome() {
        java.lang.Object ref = chromosome_;
        if (!(ref instanceof java.lang.String)) {
          com.google.protobuf.ByteString bs =
              (com.google.protobuf.ByteString) ref;
          java.lang.String s = bs.toStringUtf8();
          chromosome_ = s;
          return s;
        } else {
          return (java.lang.String) ref;
        }
      }
      /**
       * <code>optional string chromosome = 1;</code>
       */
      public com.google.protobuf.ByteString
          getChromosomeBytes() {
        java.lang.Object ref = chromosome_;
        if (ref instanceof String) {
          com.google.protobuf.ByteString b = 
              com.google.protobuf.ByteString.copyFromUtf8(
                  (java.lang.String) ref);
          chromosome_ = b;
          return b;
        } else {
          return (com.google.protobuf.ByteString) ref;
        }
      }
      /**
       * <code>optional string chromosome = 1;</code>
       */
      public Builder setChromosome(
          java.lang.String value) {
        if (value == null) {
    throw new NullPointerException();
  }
  
        chromosome_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional string chromosome = 1;</code>
       */
      public Builder clearChromosome() {
        
        chromosome_ = getDefaultInstance().getChromosome();
        onChanged();
        return this;
      }
      /**
       * <code>optional string chromosome = 1;</code>
       */
      public Builder setChromosomeBytes(
          com.google.protobuf.ByteString value) {
        if (value == null) {
    throw new NullPointerException();
  }
  checkByteStringIsUtf8(value);
        
        chromosome_ = value;
        onChanged();
        return this;
      }

      private int position_ ;
      /**
       * <code>optional uint32 position = 2;</code>
       */
      public int getPosition() {
        return position_;
      }
      /**
       * <code>optional uint32 position = 2;</code>
       */
      public Builder setPosition(int value) {
        
        position_ = value;
        onChanged();
        return this;
      }
      /**
       * <code>optional uint32 position = 2;</code>
       */
      public Builder clearPosition() {
        
        position_ = 0;
        onChanged();
        return this;
      }

      private java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord> records_ =
        java.util.Collections.emptyList();
      private void ensureRecordsIsMutable() {
        if (!((bitField0_ & 0x00000004) == 0x00000004)) {
          records_ = new java.util.ArrayList<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord>(records_);
          bitField0_ |= 0x00000004;
         }
      }

      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder> recordsBuilder_;

      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord> getRecordsList() {
        if (recordsBuilder_ == null) {
          return java.util.Collections.unmodifiableList(records_);
        } else {
          return recordsBuilder_.getMessageList();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public int getRecordsCount() {
        if (recordsBuilder_ == null) {
          return records_.size();
        } else {
          return recordsBuilder_.getCount();
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord getRecords(int index) {
        if (recordsBuilder_ == null) {
          return records_.get(index);
        } else {
          return recordsBuilder_.getMessage(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder setRecords(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord value) {
        if (recordsBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureRecordsIsMutable();
          records_.set(index, value);
          onChanged();
        } else {
          recordsBuilder_.setMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder setRecords(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder builderForValue) {
        if (recordsBuilder_ == null) {
          ensureRecordsIsMutable();
          records_.set(index, builderForValue.build());
          onChanged();
        } else {
          recordsBuilder_.setMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder addRecords(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord value) {
        if (recordsBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureRecordsIsMutable();
          records_.add(value);
          onChanged();
        } else {
          recordsBuilder_.addMessage(value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder addRecords(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord value) {
        if (recordsBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          ensureRecordsIsMutable();
          records_.add(index, value);
          onChanged();
        } else {
          recordsBuilder_.addMessage(index, value);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder addRecords(
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder builderForValue) {
        if (recordsBuilder_ == null) {
          ensureRecordsIsMutable();
          records_.add(builderForValue.build());
          onChanged();
        } else {
          recordsBuilder_.addMessage(builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder addRecords(
          int index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder builderForValue) {
        if (recordsBuilder_ == null) {
          ensureRecordsIsMutable();
          records_.add(index, builderForValue.build());
          onChanged();
        } else {
          recordsBuilder_.addMessage(index, builderForValue.build());
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder addAllRecords(
          java.lang.Iterable<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord> values) {
        if (recordsBuilder_ == null) {
          ensureRecordsIsMutable();
          com.google.protobuf.AbstractMessageLite.Builder.addAll(
              values, records_);
          onChanged();
        } else {
          recordsBuilder_.addAllMessages(values);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder clearRecords() {
        if (recordsBuilder_ == null) {
          records_ = java.util.Collections.emptyList();
          bitField0_ = (bitField0_ & ~0x00000004);
          onChanged();
        } else {
          recordsBuilder_.clear();
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public Builder removeRecords(int index) {
        if (recordsBuilder_ == null) {
          ensureRecordsIsMutable();
          records_.remove(index);
          onChanged();
        } else {
          recordsBuilder_.remove(index);
        }
        return this;
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder getRecordsBuilder(
          int index) {
        return getRecordsFieldBuilder().getBuilder(index);
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder getRecordsOrBuilder(
          int index) {
        if (recordsBuilder_ == null) {
          return records_.get(index);  } else {
          return recordsBuilder_.getMessageOrBuilder(index);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public java.util.List<? extends org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder> 
           getRecordsOrBuilderList() {
        if (recordsBuilder_ != null) {
          return recordsBuilder_.getMessageOrBuilderList();
        } else {
          return java.util.Collections.unmodifiableList(records_);
        }
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder addRecordsBuilder() {
        return getRecordsFieldBuilder().addBuilder(
            org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder addRecordsBuilder(
          int index) {
        return getRecordsFieldBuilder().addBuilder(
            index, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.getDefaultInstance());
      }
      /**
       * <code>repeated .org.opencb.biodata.models.variant.protobuf.VcfRecord records = 3;</code>
       *
       * <pre>
       * List of records (lines)
       * </pre>
       */
      public java.util.List<org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder> 
           getRecordsBuilderList() {
        return getRecordsFieldBuilder().getBuilderList();
      }
      private com.google.protobuf.RepeatedFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder> 
          getRecordsFieldBuilder() {
        if (recordsBuilder_ == null) {
          recordsBuilder_ = new com.google.protobuf.RepeatedFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecord.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfRecordOrBuilder>(
                  records_,
                  ((bitField0_ & 0x00000004) == 0x00000004),
                  getParentForChildren(),
                  isClean());
          records_ = null;
        }
        return recordsBuilder_;
      }

      private org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields fields_ = null;
      private com.google.protobuf.SingleFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.FieldsOrBuilder> fieldsBuilder_;
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public boolean hasFields() {
        return fieldsBuilder_ != null || fields_ != null;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields getFields() {
        if (fieldsBuilder_ == null) {
          return fields_ == null ? org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.getDefaultInstance() : fields_;
        } else {
          return fieldsBuilder_.getMessage();
        }
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public Builder setFields(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields value) {
        if (fieldsBuilder_ == null) {
          if (value == null) {
            throw new NullPointerException();
          }
          fields_ = value;
          onChanged();
        } else {
          fieldsBuilder_.setMessage(value);
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public Builder setFields(
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder builderForValue) {
        if (fieldsBuilder_ == null) {
          fields_ = builderForValue.build();
          onChanged();
        } else {
          fieldsBuilder_.setMessage(builderForValue.build());
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public Builder mergeFields(org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields value) {
        if (fieldsBuilder_ == null) {
          if (fields_ != null) {
            fields_ =
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.newBuilder(fields_).mergeFrom(value).buildPartial();
          } else {
            fields_ = value;
          }
          onChanged();
        } else {
          fieldsBuilder_.mergeFrom(value);
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public Builder clearFields() {
        if (fieldsBuilder_ == null) {
          fields_ = null;
          onChanged();
        } else {
          fields_ = null;
          fieldsBuilder_ = null;
        }

        return this;
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder getFieldsBuilder() {
        
        onChanged();
        return getFieldsFieldBuilder().getBuilder();
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.FieldsOrBuilder getFieldsOrBuilder() {
        if (fieldsBuilder_ != null) {
          return fieldsBuilder_.getMessageOrBuilder();
        } else {
          return fields_ == null ?
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.getDefaultInstance() : fields_;
        }
      }
      /**
       * <code>optional .org.opencb.biodata.models.variant.protobuf.Fields fields = 4;</code>
       */
      private com.google.protobuf.SingleFieldBuilder<
          org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.FieldsOrBuilder> 
          getFieldsFieldBuilder() {
        if (fieldsBuilder_ == null) {
          fieldsBuilder_ = new com.google.protobuf.SingleFieldBuilder<
              org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.Fields.Builder, org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.FieldsOrBuilder>(
                  getFields(),
                  getParentForChildren(),
                  isClean());
          fields_ = null;
        }
        return fieldsBuilder_;
      }
      public final Builder setUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }

      public final Builder mergeUnknownFields(
          final com.google.protobuf.UnknownFieldSet unknownFields) {
        return this;
      }


      // @@protoc_insertion_point(builder_scope:org.opencb.biodata.models.variant.protobuf.VcfSlice)
    }

    // @@protoc_insertion_point(class_scope:org.opencb.biodata.models.variant.protobuf.VcfSlice)
    private static final org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice DEFAULT_INSTANCE;
    static {
      DEFAULT_INSTANCE = new org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice();
    }

    public static org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice getDefaultInstance() {
      return DEFAULT_INSTANCE;
    }

    private static final com.google.protobuf.Parser<VcfSlice>
        PARSER = new com.google.protobuf.AbstractParser<VcfSlice>() {
      public VcfSlice parsePartialFrom(
          com.google.protobuf.CodedInputStream input,
          com.google.protobuf.ExtensionRegistryLite extensionRegistry)
          throws com.google.protobuf.InvalidProtocolBufferException {
        try {
          return new VcfSlice(input, extensionRegistry);
        } catch (RuntimeException e) {
          if (e.getCause() instanceof
              com.google.protobuf.InvalidProtocolBufferException) {
            throw (com.google.protobuf.InvalidProtocolBufferException)
                e.getCause();
          }
          throw e;
        }
      }
    };

    public static com.google.protobuf.Parser<VcfSlice> parser() {
      return PARSER;
    }

    @java.lang.Override
    public com.google.protobuf.Parser<VcfSlice> getParserForType() {
      return PARSER;
    }

    public org.opencb.biodata.models.variant.protobuf.VcfSliceProtos.VcfSlice getDefaultInstanceForType() {
      return DEFAULT_INSTANCE;
    }

  }

  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_Fields_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_Fields_fieldAccessorTable;
  private static com.google.protobuf.Descriptors.Descriptor
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_descriptor;
  private static
    com.google.protobuf.GeneratedMessage.FieldAccessorTable
      internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_fieldAccessorTable;

  public static com.google.protobuf.Descriptors.FileDescriptor
      getDescriptor() {
    return descriptor;
  }
  private static com.google.protobuf.Descriptors.FileDescriptor
      descriptor;
  static {
    java.lang.String[] descriptorData = {
      "\n\017VcfRecord.proto\022*org.opencb.biodata.mo" +
      "dels.variant.protobuf\032\rvariant.proto\"4\n\t" +
      "VcfSample\022\025\n\rsample_values\030\001 \003(\t\022\020\n\010gt_i" +
      "ndex\030\002 \001(\r\"\336\003\n\tVcfRecord\022\026\n\016relative_sta" +
      "rt\030\001 \001(\005\022\024\n\014relative_end\030\002 \001(\005\022\021\n\trefere" +
      "nce\030\003 \001(\t\022\021\n\talternate\030\004 \001(\t\022\017\n\007quality\030" +
      "\005 \001(\002\022E\n\004type\030\014 \001(\01627.org.opencb.biodata" +
      ".models.variant.protobuf.VariantType\022\014\n\004" +
      "call\030\r \001(\t\022\024\n\014filter_index\030\006 \001(\r\022\026\n\016id_n" +
      "on_default\030\007 \003(\t\022\032\n\016info_key_index\030\010 \003(\r",
      "B\002\020\001\022\022\n\ninfo_value\030\t \003(\t\022\023\n\013formatIndex\030" +
      "\n \001(\r\022F\n\007samples\030\013 \003(\01325.org.opencb.biod" +
      "ata.models.variant.protobuf.VcfSample\022\\\n" +
      "\023secondaryAlternates\030\016 \003(\0132?.org.opencb." +
      "biodata.models.variant.protobuf.Alternat" +
      "eCoordinate\"e\n\006Fields\022\021\n\tinfo_keys\030\001 \003(\t" +
      "\022\031\n\021default_info_keys\030\002 \003(\r\022\017\n\007filters\030\003" +
      " \003(\t\022\017\n\007formats\030\005 \003(\t\022\013\n\003gts\030\006 \003(\t\"\274\001\n\010V" +
      "cfSlice\022\022\n\nchromosome\030\001 \001(\t\022\020\n\010position\030" +
      "\002 \001(\r\022F\n\007records\030\003 \003(\01325.org.opencb.biod",
      "ata.models.variant.protobuf.VcfRecord\022B\n" +
      "\006fields\030\004 \001(\01322.org.opencb.biodata.model" +
      "s.variant.protobuf.FieldsB<\n*org.opencb." +
      "biodata.models.variant.protobufB\016VcfSlic" +
      "eProtosb\006proto3"
    };
    com.google.protobuf.Descriptors.FileDescriptor.InternalDescriptorAssigner assigner =
        new com.google.protobuf.Descriptors.FileDescriptor.    InternalDescriptorAssigner() {
          public com.google.protobuf.ExtensionRegistry assignDescriptors(
              com.google.protobuf.Descriptors.FileDescriptor root) {
            descriptor = root;
            return null;
          }
        };
    com.google.protobuf.Descriptors.FileDescriptor
      .internalBuildGeneratedFileFrom(descriptorData,
        new com.google.protobuf.Descriptors.FileDescriptor[] {
          org.opencb.biodata.models.variant.protobuf.VariantProto.getDescriptor(),
        }, assigner);
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_descriptor =
      getDescriptor().getMessageTypes().get(0);
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VcfSample_descriptor,
        new java.lang.String[] { "SampleValues", "GtIndex", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_descriptor =
      getDescriptor().getMessageTypes().get(1);
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VcfRecord_descriptor,
        new java.lang.String[] { "RelativeStart", "RelativeEnd", "Reference", "Alternate", "Quality", "Type", "Call", "FilterIndex", "IdNonDefault", "InfoKeyIndex", "InfoValue", "FormatIndex", "Samples", "SecondaryAlternates", });
    internal_static_org_opencb_biodata_models_variant_protobuf_Fields_descriptor =
      getDescriptor().getMessageTypes().get(2);
    internal_static_org_opencb_biodata_models_variant_protobuf_Fields_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_Fields_descriptor,
        new java.lang.String[] { "InfoKeys", "DefaultInfoKeys", "Filters", "Formats", "Gts", });
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_descriptor =
      getDescriptor().getMessageTypes().get(3);
    internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_fieldAccessorTable = new
      com.google.protobuf.GeneratedMessage.FieldAccessorTable(
        internal_static_org_opencb_biodata_models_variant_protobuf_VcfSlice_descriptor,
        new java.lang.String[] { "Chromosome", "Position", "Records", "Fields", });
    org.opencb.biodata.models.variant.protobuf.VariantProto.getDescriptor();
  }

  // @@protoc_insertion_point(outer_class_scope)
}
